### PR TITLE
Move put validation to commit time, update put state during validation

### PR DESCRIPTION
**Pull request by dmitrii-ubskii**
## Usage and product changes

We now avoid reading storage during put operations in a write snapshot. Instead, all put operations are initialized during commit into storage, and the state of whether the insertion of the put data needs to be performed is updated during commit validation.

### CODE REVIEW

File: common/primitive/u80.rs [hunk 1]
```diff
@@ -38,7 +38,7 @@ impl U80 {
     pub const BYTES: usize = 10;
 
     pub fn new(number: u128) -> U80 {
-        assert!(number < U80::MAX.number);
+        assert!(number <= U80::MAX.number);
```
Comment (dmitrii-ubskii) [inline common/primitive/u80.rs:41]: Minor bug. `MAX` is a valid `U80`.

File: common/primitive/u80.rs [hunk 2]
```diff
@@ -86,14 +86,14 @@ impl AddAssign<u128> for U80 {
...
     }
 }
 
 impl Sub<u128> for U80 {
     type Output = U80;
     fn sub(self, rhs: u128) -> Self::Output {
-        U80 { number: self.number - rhs }
+        Self::new(self.number - rhs)
     }
 }
```
Comment (dmitrii-ubskii) [inline common/primitive/u80.rs:98]: Making sure these ops go through the bounds check in `new()`

File: storage/isolation_manager.rs [hunk 3]
```diff
@@ -581,6 +581,54 @@ impl CommitRecord {
         // TODO: handle error with a better message
         bincode::deserialize_from(reader).unwrap_or_log()
     }
+
+    fn compute_dependency(&self, predecessor: &CommitRecord) -> CommitDependency {
```
Comment (dmitrii-ubskii) [inline storage/isolation_manager.rs:585]: Renamed and moved. No longer returns a `Result<>`, but a dependency between the two commits.

File: storage/isolation_manager.rs [hunk 4]
```diff
@@ -463,6 +446,23 @@ impl<const SIZE: usize> TimelineWindow<SIZE> {
         })
     }
 
+    fn await_pending_status_commits(&self, at: SequenceNumber) -> bool {
```
Comment (dmitrii-ubskii) [inline storage/isolation_manager.rs:449]: Moved into `TimelineWindow` with no additional changes.

File: storage/isolation_manager.rs [hunk 5]
```diff
@@ -95,89 +94,15 @@ impl IsolationManager {
...
-                } else {
-                    Ok(())
-                }
-            }
-            CommitStatus::Committed(predecessor_record) => self.validate_isolation(commit_record, predecessor_record),
-            CommitStatus::Closed => Ok(()),
-        }
-    }
 
-    fn await_pending_status_commits(
```
Comment (dmitrii-ubskii) [inline storage/isolation_manager.rs:131]: Moved into `TimelineWindow`

File: storage/isolation_manager.rs [hunk 6]
```diff
@@ -95,89 +94,15 @@ impl IsolationManager {
...
-                    // could also do a little sleep in the spin lock, for example if the validating is still far away
-                    std::hint::spin_loop();
-                }
-                CommitStatus::Committed(_) => return true,
-                CommitStatus::Closed => return false,
-            }
-        }
-    }
-
-    fn validate_isolation(&self, record: &CommitRecord, predecessor: &CommitRecord) -> Result<(), IsolationError> {
```
Comment (dmitrii-ubskii) [inline storage/isolation_manager.rs:152]: Moved into `CommitRecord`

File: storage/isolation_manager.rs [hunk 7]
```diff
@@ -95,89 +94,15 @@ impl IsolationManager {
...
                 predecessor_window = self.timeline.get_window(at);
             }
-            self.validate_concurrent(commit_record, at, &predecessor_window)?;
+            resolve_concurrent(commit_record, at, &predecessor_window)?;
             at = SequenceNumber::new(at.number() + 1);
         }
-        Ok(())
-    }
-
-    fn validate_concurrent(
```
Comment (dmitrii-ubskii) [inline storage/isolation_manager.rs:108]: Moved out

File: storage/isolation_manager.rs [hunk 8]
```diff
@@ -212,17 +137,75 @@ impl IsolationManager {
     }
 }
 
+fn resolve_concurrent(
```
Comment (dmitrii-ubskii) [inline storage/isolation_manager.rs:140]: Previously `validate_concurrent`.

Not entirely happy with this name: this procedure does a lot of things, but the main ones are conflict detection, updating the state of puts, and potentially waiting for the predecessor to finish validation unless the commits are independent.

File: storage/storage.rs [hunk 9]
```diff
@@ -273,6 +273,26 @@ impl<D> MVCCStorage<D> {
...
+                let existing_stored: Option<Option<ByteArray<BUFFER_VALUE_INLINE>>> = self
+                    .get(wrapped, commit_record.open_sequence_number(), |reference| {
+                        // Only copy if the value is the same
+                        (reference.bytes() == value.bytes()).then(|| ByteArray::from(reference))
+                    })
+                    .unwrap(); // TODO
+                reinsert.store(existing_stored.flatten().is_none(), Ordering::Release);
+            }
+        }
+
```
Comment (dmitrii-ubskii) [inline storage/storage.rs:295]: Initial Put state validation moved here

File: storage/isolation_manager.rs [hunk 10]
```diff
@@ -212,17 +137,75 @@ impl IsolationManager {
...
+        CommitDependency::Conflict(conflict) => return Err(IsolationError::Conflict(conflict)),
+    }
+
+    Ok(())
+}
+
 #[derive(Debug)]
-pub enum IsolationError {
+struct DependentPut {
+    flag: Arc<AtomicBool>,
```
Comment (flyingsilverfin) [inline storage/isolation_manager.rs:172]: I'm confused by this API of 'flag' and 'value' both being booleans, please leave a comment here

Comment (dmitrii-ubskii) [reply storage/isolation_manager.rs:172]: The `flag` is the put reinsert flag, and the value is what that flag needs to be set to. A pointer and a value.

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**